### PR TITLE
Add SpliteratorAssert

### PR DESCRIPTION
**Pull request by wjbakker**
New functionality:
```
assertThat(spliterator).hasCharacteristics(Spliterator.SIZED));
assertThat(spliterator).hasOnlyCharacteristics(Spliterator.SIZED));
```

#### Details

**Given**:
```
Spliterator<Integer> spliterator = Stream.of(1, 2, 3).spliterator();
```

**Assert hasCharacteristics**
Instead of:
```
assertThat(spliterator.hasCharacteristics(Spliterator.DISTINCT)).isTrue()
# AssertionFailedError: 
# Expecting:
#  <false>
# to be equal to:
#  <true>
# but was not.
```
one can now use:
```
assertThat(spliterator).hasCharacteristics(Spliterator.DISTINCT));
# AssertionError: 
# Expecting:
#  <["SUBSIZED", "ORDERED", "SIZED", "IMMUTABLE"]>
# to contain:
#  <["DISTINCT"]>
# but could not find:
#  <["DISTINCT"]>
```

**Assert hasOnlyCharacteristics**
Instead of
```
assertThat(spliterator.characteristics()).isEqualTo(Spliterator.SIZED | Spliterator.SUBSIZED)
# AssertionFailedError
# Expecting:
# <17488>
# to be equal to:
# <16448>
```
one can now use:
```
assertThat(spliterator).hasOnlyCharacteristics(Spliterator.SIZED));
# AssertionError: 
# Expecting:
#  <["SUBSIZED", "ORDERED", "SIZED", "IMMUTABLE"]>
# to contain only:
#  <["SUBSIZED", "SIZED"]>
# but the following elements were unexpected:
#  <["ORDERED", "IMMUTABLE"]>
```

#### Check List:
* Fixes Missing functionality
* Unit tests : YES
* Javadoc with a code example (API only) : YES

### CODE REVIEW

File: src/test/java/org/assertj/core/api/spliterator/SpliteratorAssert_hasCharacteristics_Test.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    // THEN
+    assertThat(actual).hasCharacteristics(Spliterator.DISTINCT);
+  }
+
+  @Test
+  public void should_fail_if_spliterator_does_not_have_characteristic() {
+    // GIVEN
+    Spliterator<?> actual = createSpliterator(Spliterator.SORTED);
+    // WHEN
+    Throwable thrown = catchThrowable(() -> assertThat(actual).hasCharacteristics(Spliterator.DISTINCT));
```
[the remark this answers is not in the document]
Comment (wjbakker) [reply src/test/java/org/assertj/core/api/spliterator/SpliteratorAssert_hasCharacteristics_Test.java]: Agree. Made code consistent with other units and added SpliteratorsBaseTest. Note this still relies on the ShouldContain/ShouldContainOnly error factories. I contemplated on adding ShouldHaveCharacteristics/ShouldHaveOnlyCharacteristics error factories, though its use was limited. Please let me know if you think that is useful.

File: src/test/java/org/assertj/core/api/Assertions_assertThat_with_Spliterator_Test.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  @Test
+  public void isExactlyInstanceOf_should_check_the_original_spliterator() {
+    assertThat(new StringSpliterator()).isExactlyInstanceOf(StringSpliterator.class);
+  }
+
+  @Test
+  public void isNotExactlyInstanceOf_should_check_the_original_spliterator() {
+    assertThat(stringSpliterator).isNotExactlyInstanceOf(Spliterator.class);
+    try {
+      assertThat(stringSpliterator).isNotExactlyInstanceOf(StringSpliterator.class);
```
[the remark this answers is not in the document]
Comment (wjbakker) [reply src/test/java/org/assertj/core/api/Assertions_assertThat_with_Spliterator_Test.java]: Done

File: src/test/java/org/assertj/core/api/Assertions_assertThat_with_Spliterator_Test.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    Spliterator<?> spliterator = mock(Spliterator.class);
+    assertThat(spliterator).isSameAs(spliterator);
+    verifyZeroInteractions(spliterator);
+  }
+
+  @Test
+  public void isNotSameAs_should_check_the_original_spliterator_without_consuming_it() {
+    Spliterator<?> spliterator = mock(Spliterator.class);
+    try {
+      assertThat(spliterator).isNotSameAs(spliterator);
```
[the remark this answers is not in the document]
Comment (wjbakker) [reply src/test/java/org/assertj/core/api/Assertions_assertThat_with_Spliterator_Test.java]: Done.

File: src/main/java/org/assertj/core/internal/Spliterators.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    if (hasCharacteristic(characteristics, Spliterator.ORDERED)) { names.add(SPLITERATOR_ORDERED); }
+    if (hasCharacteristic(characteristics, Spliterator.SIZED)) { names.add(SPLITERATOR_SIZED); }
+    if (hasCharacteristic(characteristics, Spliterator.NONNULL)) { names.add(SPLITERATOR_NONNULL); }
+    if (hasCharacteristic(characteristics, Spliterator.IMMUTABLE)) { names.add(SPLITERATOR_IMMUTABLE); }
+    if (hasCharacteristic(characteristics, Spliterator.CONCURRENT)) { names.add(SPLITERATOR_CONCURRENT); }
+    if (hasCharacteristic(characteristics, Spliterator.SUBSIZED)) { names.add(SPLITERATOR_SUBSIZED); }
+    return names;
+  }
+
+  private boolean hasCharacteristic(int characteristics, int characteristic) {
```
[the remark this answers is not in the document]
Comment (wjbakker) [reply src/main/java/org/assertj/core/internal/Spliterators.java]: Done (plus 2 more methods)

File: src/main/java/org/assertj/core/api/Assertions.java [hunk 5]
```diff
@@ -2759,6 +2761,23 @@ public static void useDefaultDateFormatsOnly() {
     return AssertionsForInterfaceTypes.assertThat(actual);
   }
 
+  /**
+   * Creates a new instance of <code>{@link SpliteratorAssert}</code> from the given {@link Spliterator}.
+   *
+   * Example:
+   * <pre><code class='java'>
```
[the remark this answers is not in the document]
Comment (wjbakker) [reply src/main/java/org/assertj/core/api/Assertions.java]: Done

File: src/main/java/org/assertj/core/api/AbstractSpliteratorAssert.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    AbstractAssert<SELF, Spliterator<ELEMENT>> {
+
+  private final Spliterators spliterators = Spliterators.instance();
+
+  public AbstractSpliteratorAssert(Spliterator<ELEMENT> actual, Class<?> selfType) {
+    super(actual, selfType);
+  }
+
+  /**
+   * Asserts the actual <code>{@link Spliterator}</code> has the given characteristics.
```
[the remark this answers is not in the document]
Comment (wjbakker) [reply src/main/java/org/assertj/core/api/AbstractSpliteratorAssert.java]: Done

**Comment by joel-costigliola** [comment]
Integrated thanks @wjbakker!

**Pull request closed without merging**